src/main/java/io/personium/core/ws/WebSocketService.java

```diff
@@ -223,110 +218,239 @@ public void onMessage(String text, Session session) {
+
+        if (state.equals(STATE_TYPE_ALL)
+                ||  state.equals(STATE_TYPE_SUBSCRIBE)) {
+            List<JSONObject> ruleJsonList = new ArrayList<>();
+            synchronized (lockObj) {
+                List<RuleInfo> ruleList = (List<RuleInfo>) userProperties.get(RULES);
+                for (RuleInfo rule : ruleList) {
+                    JSONObject ruleJson = new JSONObject();
+                    ruleJson.put("Type", rule.type);
+                    ruleJson.put("Object", rule.object);
+                    ruleJsonList.add(ruleJson);
+                }
+            }
+            if (state.equals(STATE_TYPE_SUBSCRIBE)) {
+                result.put(RESPONSE, STATE);
+                result.put(RESULT, RESPONSE_SUCCESS);
+                result.put("Subscriptions", ruleJsonList);
+            } else if (state.equals(STATE_TYPE_ALL)) {
+                result.put(RESPONSE, STATE);
+                result.put(RESULT, RESPONSE_SUCCESS);
+                result.put("Subscriptions", ruleJsonList);
+                Date now = new Date();
+                long expireTime = authorizedDate.getTime() + EXPIRES_IN_SECONDS * MILLISECS_IN_A_SEC;
+                long expireLimit = (expireTime - now.getTime()) / MILLISECS_IN_A_SEC;
+                result.put("ExpiresIn", expireLimit);
+                String cellName = null;
+                Cell cell = ModelFactory.cell(cellId, null);
+                if (cell != null) {
+                    cellName = cell.getName();
+                }
+                result.put("Cell", cellName);
+            }
+        } else {
+            result.put(RESPONSE, STATE);
+            result.put(RESULT, RESPONSE_ERROR);
+            result.put(REASON, REASON_INVALID_STATE_TYPE);
+        }
+        sendText(session, result.toJSONString());
+    }
+
+    /**
+     * On receive external event.
+     * @param session
+     * @param event
+     */
+    private void onReceiveExEvent(Session session, JSONObject event) {
+        log.debug("ws: " + EVENT + " : " + event.get("Type"));
+        Map<String, Object> userProperties = session.getUserProperties();
+        String cellId = (String) userProperties.get(CELL_ID);
+        String accessToken = (String) userProperties.get(ACCESS_TOKEN);
+        JSONObject result = createResultJSONObject();
+
+        try {
+            // TODO It may be better cellRsCmp is managed by HashMap
+            Cell cell = ModelFactory.cell(cellId, null);
+            AccessContext ac = createAccessContext(accessToken, cell.getName());
+            CellCmp cellCmp = ModelFactory.cellCmp(cell);
+            if (cellCmp.exists()) {
+                CellRsCmp cellRsCmp = new CellRsCmp(cellCmp, cell, ac);
+                cellRsCmp.checkAccessContext(ac, CellPrivilege.EVENT);
+                PersoniumEvent pEvent = new PersoniumEvent(
+                        PersoniumEvent.EXTERNAL_EVENT,
+                        (String) event.get("Type"),
```

Comment:
-- Why not turn these into constants ? --
Type
Object
Info
RequestKey
Schema
Subject
Timestamp
External

Reply:
Probably these constants should not rather be defined in this class.
Classes such as Event or EventBus probably have these constants already so 
we should use them instead.